include/common/ThreadPool.h

```diff
@@ -0,0 +1,75 @@
+#ifndef THREAD_POOL_H
+#define THREAD_POOL_H
+
+#include <vector>
+#include <deque>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <type_traits>
+
+namespace Common {
+    class ThreadPool {
+    private:
+        std::deque<std::function<void()>> task_queue;
+        std::vector<std::thread> workers;
+        std::mutex queue_mutex;
+        std::condition_variable worker_notification;
+        bool should_stop;
+
+        void run()
```

Comment:
I think this `run` function definition can actually be moved to the `.cpp` file since it isn't templated or wrapped in a templated class. But that's a small thing.